[5.5.x] Add since flag to gravity report

bernardjkim opened the pull request:
## Description
<!--Required. Provide high-level overview of what the change is for.-->
This PR adds the `since` flag to `gravity report` command. This flag is used to filter the log entries by time. The current implementation only affects the log entries collected from `kubectl logs` and `journalctl`.

Add a few more system collectors that have come up recently:
* Reboots history: `last -x`
* Kernel information: `uname -a`
* Serf cluster: `serf members`
* Swap status: `swapon -s (we have vmstat but wouldn't hurt)`
* Etcd metrics: `curl -s --insecure --tlsv1.2 --cacert /var/state/root.cert --cert /var/state/etcd.cert --key /var/state/etcd.key https://localhost:4001/metrics`

## Type of change
<!--Required. Keep only those that apply.-->

* New feature (non-breaking change which adds functionality)
* This change has a user-facing impact

## Linked tickets and other PRs
<!--Required. Keep only those that apply.-->

<!--This PR addresses the following issues.-->
* Updates https://github.com/gravitational/gravity/issues/1273

## TODOs
<!--Required. Keep only those that apply and check them off as they get completed.-->

- [x] Self-review the change
- [x] Perform manual testing
- [x] Address review feedback

## Testing done
<!--Required. Explain what kind of testing these changes underwent.-->
**Verify additional collectors**
```
[gravity-report] tree report                                                                                                                   bernard/5.5.x/time-filter  ✭ ✱
report
├── cluster.json
├── node-1-debug-logs
│   ├── bridge-fdb
│   ├── cluster-status
│   ├── df
│   ├── df-inodes
│   ├── dmesg
│   ├── dmsetup
│   ├── etcd-status
│   ├── etc-logs.tar.gz
│   ├── fdisk
│   ├── free
│   ├── gravity-install.log
│   ├── gravity-packages.yaml
│   ├── gravity-system.log
│   ├── gravity-system.log.gz
│   ├── host-system-failed
│   ├── host-system-jobs
│   ├── host-system-status
│   ├── ifconfig
│   ├── ip-addr
│   ├── ip-link
│   ├── ip-maddress
│   ├── ip-neighbor
│   ├── ip-netconf
│   ├── ip-ntable
│   ├── ip-route
│   ├── ip-rule
│   ├── iptables
│   ├── ip-tcp_metrics
│   ├── lsblk
│   ├── lscpu
│   ├── lsmod
│   ├── planet-journal-export.log.gz
│   ├── planet-status
│   ├── planet-system-failed
│   ├── planet-system-jobs
│   ├── planet-system-status
│   ├── reboot-history  <---
│   ├── route
│   ├── running-processes
│   ├── serf-members  <---
│   ├── slabinfo
│   ├── slabtop
│   ├── swapon <---
│   ├── uname  <---
│   └── vmstat
├── node-1-debug-logs.tar.gz
├── node-2-etcd
│   ├── etcd-backup.json
│   └── etcd-metrics  <---
├── node-2-etcd.tar.gz
[...]
```

**Verify kubectl logs are filtered by time**

* `--since 1m`
```
[vagrant@node-1 node-2-k8s-logs]$ cat k8s-logs-kube-system-gravity-site-kn59t-gravity-site
2020-06-13T09:36:30Z WARN [PROXY:AGE] Failed to create remote tunnel: dial tcp 172.28.128.101:61024: connect: connection refused, conn: <nil>. target:172.28.128.101:61024 reversetunnel/agent.go:447
2020-06-13T09:36:30Z INFO             2020/06/13 09:36:30 http: TLS handshake error from 127.0.0.1:32834: tls: internal error: failed to clone hash runtime/asm_amd64.s:1373
2020-06-13T09:36:40Z WARN [PROXY:AGE] Failed to create remote tunnel: dial tcp 172.28.128.101:61024: connect: connection refused, conn: <nil>. target:172.28.128.101:61024 reversetunnel/agent.go:447
2020-06-13T09:36:40Z INFO             2020/06/13 09:36:40 http: TLS handshake error from 127.0.0.1:32900: tls: internal error: failed to clone hash runtime/asm_amd64.s:1373
2020-06-13T09:36:40Z WARN [TELEPROXY] "Failed to renew certificate for opscenter@gravitational.io: \nERROR REPORT:\nOriginal Error: *trace.ConnectionProblemError remote error: tls: internal error\nStack Trace:\n\t/home/bernard/go/src/github.com/gravitational/gravity/vendor/github.com/gravitational/teleport/lib/httplib/httplib.go:110 github.com/gravitational/gravity/vendor/github.com/gravitational/teleport/lib/httplib.ConvertResponse\n\t/home/bernard/go/src/github.com/gravitational/gravity/vendor/github.com/gravitational/teleport/lib/auth/clt.go:211 github.com/gravitational/gravity/vendor/github.com/gravitational/teleport/lib/auth.(*Client).PostJSON\n\t/home/bernard/go/src/github.com/gravitational/gravity/vendor/github.com/gravitational/teleport/lib/auth/clt.go:1169 github.com/gravitational/gravity/vendor/github.com/gravitational/teleport/lib/auth.(*Client).GenerateUserCert\n\t/home/bernard/go/src/github.com/gravitational/gravity/lib/process/proxy.go:125 github.com/gravitational/gravity/lib/process.(*teleportProxyService).initAuthMethods.func1\n\t/home/bernard/go/src/github.com/gravitational/gravity/lib/process/proxy.go:154 github.com/gravitational/gravity/lib/process.(*teleportProxyService).initAuthMethods\n\t/usr/lib/go-1.14/src/runtime/asm_amd64.s:1373 runtime.goexit\nUser Message: remote error: tls: internal error." process/proxy.go:155
2020-06-13T09:36:50Z WARN [PROXY:AGE] Failed to create remote tunnel: dial tcp 172.28.128.101:61024: connect: connection refused, conn: <nil>. target:172.28.128.101:61024 reversetunnel/agent.go:447
2020-06-13T09:36:58Z INFO [RBAC]      Access to create user in namespace default denied to roles Node,default-implicit-role: no allow rule matched. services/role.go:1838
2020-06-13T09:37:00Z WARN [PROXY:AGE] Failed to create remote tunnel: dial tcp 172.28.128.101:61024: connect: connection refused, conn: <nil>. target:172.28.128.101:61024 reversetunnel/agent.go:447
2020-06-13T09:37:10Z WARN [PROXY:AGE] Failed to create remote tunnel: dial tcp 172.28.128.101:61024: connect: connection refused, conn: <nil>. target:172.28.128.101:61024 reversetunnel/agent.go:447
2020-06-13T09:37:20Z WARN [PROXY:AGE] Failed to create remote tunnel: dial tcp 172.28.128.101:61024: connect: connection refused, conn: <nil>. target:172.28.128.101:61024 reversetunnel/agent.go:447
```

**Verify journal logs are filtered by time**
* `--since 1m`
* Not entirely sure how to read these timestamps, but there are much fewer log entries compared to the report without a time filter provided.
```
[vagrant@node-1 node-1-debug-logs]$ cat gravity-system.log
__CURSOR=s=05ed0a8050e54d6eb25b26ebd98fd6d3;i=1d5d;b=7a3504e73fcc4cb186958464f737b949;m=1ea97c5500;t=5a7f3eea4dbd4;x=bfdc13f631df90fd
__REALTIME_TIMESTAMP=1592041006226388
__MONOTONIC_TIMESTAMP=131692516608
_BOOT_ID=7a3504e73fcc4cb186958464f737b949
_UID=0
_MACHINE_ID=ee32f6c4a34e4799afc544610c083982
PRIORITY=5
_CAP_EFFECTIVE=1fffffffff
_TRANSPORT=syslog
SYSLOG_FACILITY=10
_AUDIT_LOGINUID=1000
_SYSTEMD_OWNER_UID=1000
_SYSTEMD_SLICE=user-1000.slice
SYSLOG_IDENTIFIER=sudo
_GID=1000
_COMM=sudo
_EXE=/usr/bin/sudo
_SELINUX_CONTEXT=unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023
_HOSTNAME=node-1
MESSAGE= vagrant : TTY=pts/0 ; PWD=/home/vagrant ; USER=root ; COMMAND=/bin/gravity report --help
_CMDLINE=sudo gravity report --help
_AUDIT_SESSION=105
_SYSTEMD_CGROUP=/user.slice/user-1000.slice/session-105.scope
_SYSTEMD_SESSION=105
_SYSTEMD_UNIT=session-105.scope
_PID=17162
_SOURCE_REALTIME_TIMESTAMP=1592041006223422

__CURSOR=s=05ed0a8050e54d6eb25b26ebd98fd6d3;i=1d5e;b=7a3504e73fcc4cb186958464f737b949;m=1ea97c80c1;t=5a7f3eea50796;x=6c580e6969270a4f
__REALTIME_TIMESTAMP=1592041006237590
__MONOTONIC_TIMESTAMP=131692527809
_BOOT_ID=7a3504e73fcc4cb186958464f737b949
PRIORITY=6
_UID=0
_GID=0
_MACHINE_ID=ee32f6c4a34e4799afc544610c083982
_CAP_EFFECTIVE=1fffffffff
_TRANSPORT=syslog
SYSLOG_FACILITY=10
_AUDIT_LOGINUID=1000
_SYSTEMD_OWNER_UID=1000
_SYSTEMD_SLICE=user-1000.slice
SYSLOG_IDENTIFIER=sudo
_COMM=sudo
_EXE=/usr/bin/sudo
_SELINUX_CONTEXT=unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023
_HOSTNAME=node-1
MESSAGE=pam_unix(sudo:session): session opened for user root by vagrant(uid=0)
_CMDLINE=sudo gravity report --help
_AUDIT_SESSION=105
_SYSTEMD_CGROUP=/user.slice/user-1000.slice/session-105.scope
_SYSTEMD_SESSION=105
_SYSTEMD_UNIT=session-105.scope
_PID=17162
_SOURCE_REALTIME_TIMESTAMP=1592041006236963

__CURSOR=s=05ed0a8050e54d6eb25b26ebd98fd6d3;i=1d5f;b=7a3504e73fcc4cb186958464f737b949;m=1ea97d9723;t=5a7f3eea61df7;x=db7124e0e9f6124
__REALTIME_TIMESTAMP=1592041006308855
__MONOTONIC_TIMESTAMP=131692599075
_BOOT_ID=7a3504e73fcc4cb186958464f737b949
PRIORITY=6
_UID=0
_GID=0
_MACHINE_ID=ee32f6c4a34e4799afc544610c083982
_CAP_EFFECTIVE=1fffffffff
_TRANSPORT=syslog
SYSLOG_FACILITY=10
_AUDIT_LOGINUID=1000
_SYSTEMD_OWNER_UID=1000
_SYSTEMD_SLICE=user-1000.slice
SYSLOG_IDENTIFIER=sudo
_COMM=sudo
_EXE=/usr/bin/sudo
_SELINUX_CONTEXT=unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023
_HOSTNAME=node-1
MESSAGE=pam_unix(sudo:session): session closed for user root
_CMDLINE=sudo gravity report --help
_AUDIT_SESSION=105
_SYSTEMD_CGROUP=/user.slice/user-1000.slice/session-105.scope
_SYSTEMD_SESSION=105
_SYSTEMD_UNIT=session-105.scope
_PID=17162
_SOURCE_REALTIME_TIMESTAMP=1592041006308612

__CURSOR=s=05ed0a8050e54d6eb25b26ebd98fd6d3;i=1d60;b=7a3504e73fcc4cb186958464f737b949;m=1eac0643e1;t=5a7f3f12ecab5;x=8661f252df6da835
__REALTIME_TIMESTAMP=1592041048820405
__MONOTONIC_TIMESTAMP=131735110625
_BOOT_ID=7a3504e73fcc4cb186958464f737b949
PRIORITY=6
_UID=0
_GID=0
_SYSTEMD_SLICE=system.slice
_MACHINE_ID=ee32f6c4a34e4799afc544610c083982
SYSLOG_FACILITY=3
_CAP_EFFECTIVE=1fffffffff
_TRANSPORT=stdout
_SELINUX_CONTEXT=system_u:system_r:init_t:s0
_HOSTNAME=node-1
SYSLOG_IDENTIFIER=gravity
_COMM=teleport
_CMDLINE=rootfs/usr/bin/teleport start
_STREAM_ID=a4ff5bb0a2e34ea39c31b31ec84dbc20
_PID=1412
_EXE=/var/lib/gravity/local/packages/unpacked/gravitational.io/teleport/3.0.5/rootfs/usr/bin/teleport
_SYSTEMD_CGROUP=/system.slice/gravity__gravitational.io__teleport__3.0.5.service
_SYSTEMD_UNIT=gravity__gravitational.io__teleport__3.0.5.service
MESSAGE=INFO [NODE]      [LOCAL EXEC] Started command: "gravity --insecure system report --filter system --compressed --since 1m0s" id:9 local:172.28.128.101:3022 login:root remote:127.0.0.1:56576 teleportUser:opscenter@gravitational.io srv/exec.go:171

```

bernardjkim committed:
Add since flag to gravity report

r0mant commented on lib/ops/opsservice/report.go:
```diff
@@ -224,7 +228,9 @@ func (s *site) collectKubernetesInfo(reportWriter report.FileWriter, runner *ser
-		fmt.Sprintf("--filter=%v", report.FilterKubernetes), "--compressed")...)
+		fmt.Sprintf("--filter=%v", report.FilterKubernetes),
+		"--compressed",
+		fmt.Sprintf(`--since="%v"`, since.String()))...)
```
Same nit as above.

r0mant commented on lib/report/k8s.go:
```diff
@@ -78,7 +80,7 @@ func capturePreviousContainerLogs(ctx context.Context, namespaces []string, runn
 				name := fmt.Sprintf("k8s-logs-%v-%v-%v-prev", namespace, pod, container)
 				collectors = append(collectors, Cmd(name, kubectl.Command("logs", pod,
-					"--namespace", namespace, "-p",
+					"--namespace", namespace, "-p", "--since", since.String(),
```
This only filters the k8s logs for _previous_ instances of containers (`-p` flag). There's a cluster-info call above that still collects all logs for all currently running containers. To filter those too, maybe we can skip collecting "cluster-info" if time filter is provided, and instead use the same "kubectl logs" command as here but without "-p" to collect logs of current containers with proper time filter applied?

bernardjkim replied:
Kubernetes report structure. `kubectl cluster-info dump` collects a bit more than just the logs of each pod. Do we want to manually collect this information as well when the `--since` flag is provided?

* `--since` not provided
```
├── default
│   ├── daemonsets.json
│   ├── deployments.json
│   ├── events.json
│   ├── pods.json
│   ├── replicasets.json
│   ├── replication-controllers.json
│   └── services.json
├── k8s-logs-kube-system-coredns-9rr89-coredns-prev
├── k8s-logs-kube-system-coredns-g9xv5-coredns-prev
├── k8s-logs-kube-system-coredns-sthwk-coredns-prev
├── k8s-logs-kube-system-site-app-post-install-2bcd3f-79kft-post-install-hook-prev
├── k8s-logs-monitoring-influxdb-78f4754b8f-zql59-watcher-prev
├── kube-node-lease
│   ├── daemonsets.json
│   ├── deployments.json
│   ├── events.json
│   ├── pods.json
│   ├── replicasets.json
│   ├── replication-controllers.json
│   └── services.json
├── kube-public
│   ├── daemonsets.json
│   ├── deployments.json
│   ├── events.json
│   ├── pods.json
│   ├── replicasets.json
│   ├── replication-controllers.json
│   └── services.json
├── kube-system
│   ├── bandwagon-6c4b5b5c76-tf9kc
│   │   └── logs.txt
│   ├── bandwagon-install-ab9d29-779nz
│   │   └── logs.txt
│   ├── coredns-9rr89
│   │   └── logs.txt
│   ├── coredns-g9xv5
│   │   └── logs.txt
│   ├── coredns-sthwk
│   │   └── logs.txt
│   ├── daemonsets.json
│   ├── deployments.json
│   ├── dns-app-install-c4f31f-j2cp5
│   │   └── logs.txt
│   ├── events.json
│   ├── gravity-install-fa4102-hkpvv
│   │   └── logs.txt
│   ├── gravity-site-c7jxz
│   │   └── logs.txt
│   ├── gravity-site-xlrwl
│   │   └── logs.txt
│   ├── gravity-site-zg7j5
│   │   └── logs.txt
│   ├── log-collector-697d94486-vftvx
│   │   └── logs.txt
│   ├── log-forwarder-5rqnm
│   │   └── logs.txt
│   ├── log-forwarder-g7pk2
│   │   └── logs.txt
│   ├── log-forwarder-vbt9r
│   │   └── logs.txt
│   ├── logging-app-bootstrap-fec658-qr7q4
│   │   └── logs.txt
│   ├── monitoring-app-install-f4a4b8-mckh4
│   │   └── logs.txt
│   ├── pods.json
│   ├── replicasets.json
│   ├── replication-controllers.json
│   ├── services.json
│   ├── site-app-post-install-2bcd3f-79kft
│   │   └── logs.txt
│   ├── tiller-app-bootstrap-cb1643-7bdwt
│   │   └── logs.txt
│   └── tiller-deploy-7997c85989-9f856
│       └── logs.txt
├── monitoring
│   ├── daemonsets.json
│   ├── deployments.json
│   ├── events.json
│   ├── grafana-847888c644-vj8qz
│   │   └── logs.txt
│   ├── heapster-5db74f4bbd-5t6qx
│   │   └── logs.txt
│   ├── influxdb-78f4754b8f-zql59
│   │   └── logs.txt
│   ├── kapacitor-84c57d6cb-qpkt2
│   │   └── logs.txt
│   ├── nethealth-9dg7b
│   │   └── logs.txt
│   ├── nethealth-fz7pd
│   │   └── logs.txt
│   ├── nethealth-xgrkm
│   │   └── logs.txt
│   ├── pods.json
│   ├── replicasets.json
│   ├── replication-controllers.json
│   ├── services.json
│   ├── telegraf-7bfc97698c-s6rc6
│   │   └── logs.txt
│   ├── telegraf-node-master-rsm44
│   │   └── logs.txt
│   ├── telegraf-node-master-s7gq8
│   │   └── logs.txt
│   └── telegraf-node-master-xtzfn
│       └── logs.txt
└── nodes.json
[...]
```

* --since=1m
```
├── k8s-logs-kube-system-gravity-site-c7jxz-gravity-site
├── k8s-logs-kube-system-gravity-site-xlrwl-gravity-site
├── k8s-logs-kube-system-gravity-site-zg7j5-gravity-site
├── k8s-logs-kube-system-site-app-post-install-2bcd3f-79kft-post-install-hook-prev
├── k8s-logs-monitoring-influxdb-78f4754b8f-zql59-influxdb
├── k8s-logs-monitoring-kapacitor-84c57d6cb-qpkt2-kapacitor
[...]
```

r0mant replied:
Besides the logs, I think it also collects the dump of all resources? IIRC we already collect those too (kubectl describe outputs), so I don't think we need to do anything separately.

r0mant reviewed (commented):
Also, if we could roll a few of those additional system collectors from https://github.com/gravitational/gravity/issues/1273 in the same PR, that'd be cool (less forward-porting burden later).

bernardjkim committed:
Simplify --since arg

a-palchikov commented on lib/ops/opsservice/report.go:
```diff
@@ -208,14 +211,15 @@ func (s *site) collectDebugInfo(reportWriter report.FileWriter, runner *serverRu
 		fmt.Sprintf("--filter=%v", report.FilterSystem),
-		"--compressed")...)
+		"--compressed",
+		fmt.Sprintf(`--since="%v"`, since.String()))...)
```
or better yet:
```go
	"--compressed",
	"--since", since.String())...)
```

bernardjkim committed:
Wrap key and since in GetSiteReportRequest

bernardjkim committed:
Collect kubectl logs if since is provided

r0mant commented on lib/ops/ops.go:
```diff
@@ -1901,3 +1901,11 @@ type Identity interface {
 }
+
+// GetSiteReportRequest specifies the request to get the site report
+type GetSiteReportRequest struct {
```
Nit: I would call it `GetClusterReportRequest`. "Site" is an older term (despite still being used a lot in the code due to legacy reasons), we use "Cluster" everywhere now.

r0mant commented on lib/ops/ops.go:
```diff
@@ -1901,3 +1901,11 @@ type Identity interface {
+	// SiteKey is a key used to identify site
+	SiteKey
+	// Since is used to filter the collected report by time
+	Since time.Duration `json:"since"`
```
`json:"since,omitempty"`?

r0mant commented on lib/ops/ops.go:
```diff
@@ -1901,3 +1901,11 @@ type Identity interface {
+type GetSiteReportRequest struct {
+	// SiteKey is a key used to identify site
+	SiteKey
+	// Since is used to filter the collected report by time
```
Nit: To filter collected logs by time

r0mant commented on lib/report/k8s.go:
```diff
@@ -75,17 +100,28 @@ func capturePreviousContainerLogs(ctx context.Context, namespaces []string, runn
+	}
+	return Cmd(name, kubectl.Command("logs", pod,
+		"--namespace", namespace,
+		"--since", since.String(),
```
What if "since" is 0? Will it work as expected, i.e. collect all logs?

bernardjkim replied:
Yup, default is `0s`

`--since=0s: Only return logs newer than a relative duration like 5s, 2m, or 3h. Defaults to all logs. Only one of
since-time / since may be used.`

r0mant commented on lib/ops/opsservice/report.go:
```diff
@@ -224,7 +228,9 @@ func (s *site) collectKubernetesInfo(reportWriter report.FileWriter, runner *ser
-		fmt.Sprintf("--filter=%v", report.FilterKubernetes), "--compressed")...)
+		fmt.Sprintf("--filter=%v", report.FilterKubernetes),
+		"--compressed",
+		"--since", since.String())...)
```
I wonder if it'd make sense to only append this flag if "since" is not 0. Same above.

bernardjkim replied:
We can do that just to be safe.

r0mant commented on lib/report/system.go:
```diff
@@ -99,10 +100,14 @@ func systemStatus() Collectors {
 #!/bin/bash
-/bin/journalctl --no-pager --output=export | /bin/gzip -f`
+/bin/journalctl --no-pager --output=export `
+	if since != time.Duration(0) {
```
Nit: Wouldn't just `if since != 0` work?

r0mant commented on tool/gravity/cli/register.go:
```diff
@@ -535,6 +535,7 @@ func RegisterCommands(app *kingpin.Application) *Application {
 	// get cluster diagnostics report
 	g.ReportCmd.CmdClause = g.Command("report", "Generate cluster diagnostics report")
 	g.ReportCmd.FilePath = g.ReportCmd.Flag("file", "target report file name").Default("report.tar.gz").String()
+	g.ReportCmd.Since = g.ReportCmd.Flag("since", "only return logs newer than a relative duration like 5s, 2m, or 3h").Duration()
```
Nit: Could you please capitalize flag description sentences in these commands?

r0mant commented on lib/report/k8s.go:
```diff
@@ -50,14 +50,39 @@ func NewKubernetesCollector(ctx context.Context, runner utils.CommandRunner) Col
+	commands = append(commands, capturePreviousContainerLogs(ctx, namespaces, runner, since)...)
+
+	// collect current container logs
+	if since == time.Duration(0) {
```
Nit: `since == 0`?

r0mant commented on lib/report/report.go:
```diff
@@ -102,9 +103,18 @@ type Config struct {
+	Since time.Duration
+	// Until specifies the end of the time filter. A value of 1h will report
+	// log entries from the start of the time filter up till one hour ago.
+	Until time.Duration
```
Maybe clarify it's not being used currently and add a TODO item, or just remove completely for now so we don't keep dead code.

r0mant reviewed (commented):
Mostly cosmetic feedback.

bernardjkim committed:
GetSiteReportRequest -> GetClusterReportRequest

bernardjkim committed:
Append since flag if not 0

bernardjkim committed:
Capitalize descriptions; Remove unused field

bernardjkim committed:
Add additional collectors

- Reboots history
- Kernel information
- Serf cluster
- Swap status
- Etcd metrics

r0mant commented on lib/ops/opshandler/opshandler.go:
```diff
@@ -950,7 +950,18 @@ func (h *WebHandler) activateSite(w http.ResponseWriter, r *http.Request, p http
 func (h *WebHandler) getSiteReport(w http.ResponseWriter, r *http.Request, p httprouter.Params, context *HandlerContext) error {
-	report, err := context.Operator.GetSiteReport(siteKey(p))
+	var since time.Duration
+	if keys, ok := r.URL.Query()["since"]; ok && len(keys[0]) > 0 {
```
`if r.URL.Query().Get("since") != ""` might be a little cleaner (other handles do that).

r0mant commented on lib/webapi/webapi.go:
```diff
@@ -1696,9 +1696,20 @@ func (m *Handler) getSiteEndpoints(w http.ResponseWriter, r *http.Request, p htt
-		AccountID:  context.User.GetAccountID(),
-		SiteDomain: p.ByName("domain"),
+	var since time.Duration
+	if keys, ok := r.URL.Query()["since"]; ok && len(keys[0]) > 0 {
```
Same as above, `r.URL.Query().Get("since")`.

bernardjkim committed:
Use URL.Query().Get()

bernardjkim committed:
Edit etcd naming

bernardjkim committed:
gravity-system.log.gz -> gravity-journal.log.gz

There is already a gravity-system file. Conflict occurs when trying to
unzip gravity-system.log.gz.

knisbet commented on lib/report/report.go:
```diff
@@ -42,12 +43,13 @@ func Collect(ctx context.Context, config Config, w io.Writer) error {
+			collectors = append(collectors, NewKubernetesCollector(ctx, utils.Runner, config.Since)...)
 		case FilterEtcd:
 			collectors = append(collectors, etcdBackup()...)
+			collectors = append(collectors, etcdMetrics()...)
```
I'm not familiar enough with the filters, but based on your sample output, this seems to only collect metrics from one node in the cluster. Some of the metrics will be relevant to the particular server though, so this should likely be collected on each master node, so each server can be inspected individually. Be careful not to collect on the worker nodes, as those connections will be randomly load balanced to a master.

Unfortunately what makes this even more complicated, is many of the metrics are not useful on their own with a single snapshot. Many metrics will be counters since startup, which don't say anything on their own. It's only when you delta the current snapshot, with one a minute ago, to you get a picture of what was happening over the last minute. This isn't really just isolated to etcd, I don't think we've really planned on how to pull any metric histories as part of the debug reporting. So we might want to put together a plan on pulling out valuable metrics histories.

bernardjkim replied:
Updated so that etcd metrics are collected on all master nodes fab170f.

We should definitely look to do some additional processing of the metrics. For this PR I think the single snapshot can still provide some useful information?

bernardjkim committed:
Edit etcd metics collector

- Fetch metrics from :2379 instead of :4001
- No need to insecure flag

bernardjkim commented on tool/gravity/cli/register.go:
```diff
@@ -534,7 +534,8 @@ func RegisterCommands(app *kingpin.Application) *Application {
 	g.ReportCmd.CmdClause = g.Command("report", "Generate cluster diagnostics report")
-	g.ReportCmd.FilePath = g.ReportCmd.Flag("file", "target report file name").Default("report.tar.gz").String()
+	g.ReportCmd.FilePath = g.ReportCmd.Flag("file", "Target report file name").Default("report.tar.gz").String()
+	g.ReportCmd.Since = g.ReportCmd.Flag("since", "Only return logs newer than a relative duration like 5s, 2m, or 3h").Duration()
```
Yeah, that's a good point. We can set it to 14 days. I'll edit the description to include the default and to pass `0s` if they would like the full logs.

bernardjkim committed:
Specify 14 day default to since flag

bernardjkim committed:
Typo

bernardjkim committed:
Collect etcd metrics from all master nodes

a-palchikov commented:
Sorry I sort of missed on this PR - why was the kubernetes logs collector removed?

a-palchikov commented:
Also, the new flag is not propagated to planet journal log collector.
